### PR TITLE
Make canopen_motor_node reusable

### LINKED ISSUE #101: Make canopen_motor_node accessible via headers
Move declarations and templates into dedicated headers.

### PR DESCRIPTION
**Pull request by mistoll**
There are two changes here:

https://github.com/mistoll/ros_canopen/commit/1de23dc1807473419dbe5c7bef12e1e1f15cf0a4
This commit does only split the headers and cpp files into multiple units. 
It's mostly cut & paste. However the classes were moved to the canopen namespace.
Also a library and exports were added to CMakeLists.txt
Closes #101 

https://github.com/mistoll/ros_canopen/commit/2c774e36183cd75defd7b310900e978172056c7f
The previous commit allows using the code as a library in other packages. However the tight coupling between HandleLayer and RobotLayer reduces the cases where it is resonable to do so. Thus a pure virtual class HandleLayerBase was introduced. That allows the implementation of custom HandleLayers, which do not use the MotorBase class.
This is just a proposal. I chose this solution because it's very simple and has low impact. However there might be way better ideas, especially with the view on the whole project.

### CODE REVIEW

**Comment by mathias-luedtke** [comment]
Thanks a lot!
This will take some time to review.

> Thus a pure virtual class HandleLayerBase was introduced.

Sound reasonable :)

File: canopen_motor_node/CMakeLists.txt [hunk 1]
```diff
@@ -22,41 +22,76 @@ find_package(Boost REQUIRED
...
 # canopen_motor_node
 add_executable(canopen_motor_node
-  src/control_node.cpp
-  src/robot_layer.cpp
+  src/canopen_motor_chain_node.cpp
 )
 target_link_libraries(canopen_motor_node
+  canopen_motor
   ${catkin_LIBRARIES}
-  ${MUPARSER_LIBRARIES}
```
Comment (mathias-luedtke) [inline canopen_motor_node/CMakeLists.txt:40]: why did you remove MUPARSER here, but kept catkin?

File: canopen_motor_node/src/interface_mapping.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
+
```
Comment (mathias-luedtke) [inline canopen_motor_node/src/interface_mapping.h:1]: I am not sure if this should be part of the public API

**Comment by mathias-luedtke** [comment]
Just some minor questions :)

Great job :+1: 

Next step: I will test this with ~~kinetic~~jade

File: canopen_motor_node/CMakeLists.txt (hunk 1 above)
Comment (mistoll) [reply canopen_motor_node/CMakeLists.txt:40]: I've splitted canopen_motor_node into canopen_motor library and canopen_motor_node executable. MUPARSER is linked into canopen_motor library. 
IMO the executable canopen_motor_node does not require MUPARSER.

File: canopen_motor_node/src/interface_mapping.h (hunk 2 above)
Comment (mistoll) [reply canopen_motor_node/src/interface_mapping.h:1]: It's not. The header is placed into the src folder, not the include folder.
I've used that method for private headers in some packages and it works great. However I haven't seen that in the ros_canopen packages, yet. It might not fit into the coding guidelines.

Comment (mathias-luedtke) [reply canopen_motor_node/src/interface_mapping.h:1]: Ah, that's fine! sorry for the noise!

> I've used that method for private headers in some packages and it works great

Me too :)

File: canopen_motor_node/CMakeLists.txt (hunk 1 above)
Comment (mathias-luedtke) [reply canopen_motor_node/CMakeLists.txt:40]: I just realised that canopen_motor_node might need additional lib froms catkin, that have been stripped from canopen_motor.

**Review by mathias-luedtke — APPROVED** [review]
LGTM

I will run a prerelease test, but I don't expect any problems.

Will get merged after a hardware test.

**Comment by mathias-luedtke** [comment]
I have run a small test, everything worked as expected :+1: 
I will trigger a release.
